### ISSUE TITLE
Code generator mistake results on compile time error while injecting static nested classes

### ISSUE THREAD

**Issue by lrn2prgrm** [stephanenicolas/toothpick#181]
I want to avoid the use of Toothpick class everywhere i want an injection (if there is another way to accomplish this please let me know) so the idea was to call the injection in the constructor so it will be executed on child's super() call.

_Current behavior errors at compile time with message on Child$$MemberInjector 
cannot find symbol. symbol: Parent$$MemberInjector_
```java
    public class Child extends Parent{
      @Inject
      SomeDependency someDependency;      
      Child () {  
        super(); 
        /* I pretend injection of someDependency to happen inside super
        * without having to call and import the Toothpick class in the child class */
      }
    }
    public class Parent{
      @Inject
      OtherDependency otherDependency;      
      public RESTController() {
        Scope scope = Toothpick.openScope("scope");
        Toothpick.inject(this, scope);
      }
    }
```
I would like to suggest this as a feature request if it was possible to implement.

**Comment by stephanenicolas** [comment]
The pattern you mention is one of the possible way to use TP. I am not sure
to understand exactly what feature you request, it should just work.

Also, it's surprising that you get the error

*Child$$MemberInjectorcannot find symbol. symbol: Parent$$MemberInjector*

If Parent is accessible when compiling Child, then the Parent
MemberInjector is also accessible to the compiler when it compiles the
Child MemberInjector. I can't think of any case where this error can happen
? Can you set up a small project on GH to demonstrate the issue ?



2017-01-02 1:37 GMT+01:00 J Armando Cordova <notifications@github.com>:

**Comment by lrn2prgrm** [comment]
I haven't been able to reproduce the behavior but this code never gets injected. (both dependencies print null)

`import toothpick.Scope;
import toothpick.Toothpick;
import toothpick.config.Module;

import javax.inject.Inject;

/**
 * Created by jarma on 1/2/2017.
 */
public class TPIssueMin {
	static String scope = "scope";

	public static class Dependency {@Inject Dependency() {}}
	public static class Parent {
		@Inject	Dependency parentDependency;
		Parent() { Toothpick.inject(this, Toothpick.openScope(scope)); }
	}
	public static class Child extends Parent {
		@Inject Dependency childDependency;
		Child() { super(); }
	}

	public static void main(String[] args) {
		Scope s = Toothpick.openScope(scope);
		s.installModules(new Module() {{
			bind(Dependency.class).toInstance(new Dependency());
		}});
		Child child = new Child();
		System.out.println("childDependency: " + child.childDependency);
		System.out.println("parentDependency: " + child.parentDependency);
	}

}`

**Comment by stephanenicolas** [comment]
So, it looks like you are reporting 2 issues, right ? 
1. The first one being that the dependencies are null with the `TPIssueMin` class. 
2. The second one being the compile error you get with the `PoemController` class.

If this is the case, can you open a clear separate issue for each issue and give a clear example to reproduce it in both cases. Also, it would be nice if you could output the generated code (for the compiler error bug).

Thanks for reporting these problems. It's strange, this all should work well, we will investigate quickly.

**Comment by lrn2prgrm** [comment]
More findings: In generated class 
 TPIssue$Child$$MemberInjector calls the constructor:
new org.branchart.poetree.first.main.TPIssue.Parent$$MemberInjector();
But the name of such class is  TPIssue$Parent$$MemberInjector

It seems it boilds down to some issues with static classes/nested classes.
But the bug is weird. This problem $ character problem happens at my original proyect and in another proyect i made i only get null on the dependencies (Both running same TPIssueMin class)

**Comment by lrn2prgrm** [comment]
Error:(8, 92) java: cannot find symbol
  symbol:   class Parent$$MemberInjector
  location: class org.branchart.poetree.first.main.TPIssue

Generated class logic will join outer and inner class with a dollar TPIssue$Parent$$MemberInjector. But on injection it is calling Parent$$MemberInjector.

I haven't been able to check on inner classes (non static)

**Comment by stephanenicolas** [comment]
I was wondering about the inner class issue. So, a simple work around is to
make your classes not nested anymore. Though this is a bug and should be
fix. Thanks for reporting it !

2017-01-03 7:49 GMT+01:00 J Armando Cordova <notifications@github.com>:

**Comment by lrn2prgrm** [comment]
Yeap, disregard my "dependencies are null" I am just so silly I didn't know I had to turn on annotation processing in the new project. (I had turned on because Lombok asked me to do so. It would be nice to have it on read me for those who don't know)

**Comment by lrn2prgrm** [comment]
This is the output of the wrongly generated class:
The correct name would be: TPIssueMin$Parent$$MemberInjector
```java
import java.lang.Override;
import toothpick.MemberInjector;
import toothpick.Scope;

public final class TPIssueMin$Child$$MemberInjector implements MemberInjector<TPIssueMin.Child> {
  private MemberInjector superMemberInjector = new jarmandocordova.issue.min.TPIssueMin.Parent$$MemberInjector();

  @Override
  public void inject(TPIssueMin.Child target, Scope scope) {
    superMemberInjector.inject(target, scope);
    target.childDependency = scope.getInstance(TPIssueMin.Dependency.class);
  }
}
```

**Comment by dlemures** [comment]
PR: https://github.com/stephanenicolas/toothpick/pull/183

**Comment by dlemures** [comment]
@lrn2prgrm thx for the details!

**Comment by dlemures** [comment]
PR merged, releasing fix in 1.0.4
